runtime/gc_vlhgc/CopyForwardCompactGroup.hpp

```diff
@@ -126,12 +132,122 @@ class MM_CopyForwardCompactGroup
 	{
 		_TLHRemainderBase = NULL;
 		_TLHRemainderTop = NULL;
+		_gcIDForTLHRemainder = 0;
 	}
 
 	MMINLINE void setTLHRemainder(void *base, void *top)
 	{
 		_TLHRemainderBase = base;
 		_TLHRemainderTop = top;
+		_TLHRemainderCount += 1;
+	}
+
+	MMINLINE void moveTLHRemainder(MM_CopyForwardCompactGroup *src)
+	{
+		_TLHRemainderBase = src->_TLHRemainderBase;
+		src->_TLHRemainderBase = NULL;
+		_TLHRemainderTop = src->_TLHRemainderTop;
+		src->_TLHRemainderTop = NULL;
+		_gcIDForTLHRemainder = src->_gcIDForTLHRemainder;
+		src->_gcIDForTLHRemainder = 0;
+	}
+
+	MMINLINE uintptr_t getTLHRemainderSize()
+	{
+		return ((uintptr_t)_TLHRemainderTop - (uintptr_t)_TLHRemainderBase);
+	}
+
+	/**
+	 * return true if TLHRemainder has been aligned.
+	 */
+	MMINLINE bool alignTLHRemainder(uintptr_t tlhSurvivorDiscardThreshold)
+	{
+		bool aligned = false;
+		if (_TLHRemainderBase) {
+			void *newAlignedTLHRemainderBase = (void*) MM_Math::roundToCeilingCard((uintptr_t) _TLHRemainderBase);
+			void *newAlignedTLHRemainderTop =  (void*) MM_Math::roundToFloorCard((uintptr_t) _TLHRemainderTop);
+			if (((uintptr_t)newAlignedTLHRemainderTop - (uintptr_t)newAlignedTLHRemainderBase) >= tlhSurvivorDiscardThreshold) {
+				/* TLHRemainder need to be card aligned for being reuse cross PGCs */
+				_TLHRemainderBase = newAlignedTLHRemainderBase;
+				_TLHRemainderTop = newAlignedTLHRemainderTop;
+				aligned = true;
+			}
+			/* else aligned TLHRemainder is smaller than tlhSurvivorDiscardThreshold, need to be reset */
+		}
+		return aligned;
+	}
+
+	MMINLINE bool abandonTLHRemainder(MM_EnvironmentVLHGC* env, uintptr_t compactGroup, bool preserveRemainder)
+	{
+		bool abandoned = true;
+		if (NULL != _TLHRemainderBase) {
+			Assert_MM_true(NULL != _TLHRemainderTop);
+			/* make it a walkable hole */
+			env->_cycleState->_activeSubSpace->abandonHeapChunk(_TLHRemainderBase, _TLHRemainderTop);
+			MM_GCExtensions* extensions =MM_GCExtensions::getExtensions(env);
+			MM_HeapRegionDescriptorVLHGC *region = (MM_HeapRegionDescriptorVLHGC*)extensions->heapRegionManager->tableDescriptorForAddress(_TLHRemainderBase);
+
+			/* only preserve for non Nursery regions */
+			if (!preserveRemainder || (MM_CompactGroupManager::getRegionAgeFromGroup(env, compactGroup) < extensions->tarokNurseryMaxAge._valueSpecified) || !region->containsObjects()) {
+				/* discard the remainder */
+				discardHeapChunk(_TLHRemainderBase, _TLHRemainderTop, region);
+				resetTLHRemainder();
+
+			} else {
+				_gcIDForTLHRemainder = extensions->globalVLHGCStats.gcCount;
+				abandoned = false;
```

Comment:
I guess, this is where we could call something like ageTLHRemainder(env, compactGroup);

Reply:
sorry, I was not clear enough... I meant, you should try to move all the aging logic from abandonTLHRemainders (movement to next age group, handling max age spacial case) into ageTLHRemainder

Reply:
ok, I will try to move  aging logic from abandonTLHRemainders to ageTLHRemainder.

Reply:
I'm being picky now, but I think I'd prefer calling ageTLHRemainder from here (and then this method does not even need a return value).